### PR TITLE
feat: use env var instead of plain text for vault token

### PR DESCRIPTION
**Pull request by shreemaan-abhishek**
### Description

Currently, we use plain text to add `vault` as a secret resource. Like so:

```bash
curl http://127.0.0.1:9180/apisix/admin/secret/vault/1 \
-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "https://127.0.0.1:8200"，
    "prefix": "apisix",
    "token": "root"
}'
```

This PR implements a feature that would allow the use of environment variables in the following format: `$ENV://NAME_OF_ENV_VAR`.

This is the final outcome:

```bash
curl http://127.0.0.1:9180/apisix/admin/secret/vault/1 \
-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "https://127.0.0.1:8200"，
    "prefix": "apisix",
    "token": "$ENV://VAULT_TOKEN"
}'
```

A part of #8319 

cc: @soulbird 

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: t/config-center-yaml/secret.t [hunk 1]
```diff
@@ -61,7 +61,7 @@ __DATA__
 secrets:
   - id: vault/1
     prefix: kv/apisix
-    token: root
+    token: "$ENV://VAULT_TOKEN"
```
Comment (soulbird) [inline t/config-center-yaml/secret.t:64]: So we can no longer use such a configuration?
```
token: root
```

Comment (shreemaan-abhishek) [reply t/config-center-yaml/secret.t:64]: Yes, we won't be able to use such a configuration.

File: ci/centos7-ci.sh [hunk 2]
```diff
@@ -84,6 +84,7 @@ run_case() {
     export_or_prefix
     make init
     set_coredns
+    export VAULT_TOKEN="root"
```
Comment (soulbird) [inline ci/centos7-ci.sh:87]: refer: https://github.com/apache/apisix/blob/master/t/core/env.t#L17-L20

File: t/config-center-yaml/secret.t [hunk 3]
```diff
@@ -388,3 +392,131 @@ env secret=apisix;
...
+            local values = secret.secrets()
+            ngx.say(#values)
+        }
+    }
+--- request
+GET /t
+--- response_body
+0
+--- error_log
+property "uri" validation failed: failed to match pattern "^[^\\/]+:\\/\\/([\\da-zA-Z.-]+|\\[[\\da-fA-F:]+\\])(:\\d+)?"
```
Comment (soulbird) [inline t/config-center-yaml/secret.t:419]: Why add this test case? It doesn't look like it has anything to do with your modification？

File: t/config-center-yaml/secret.t [hunk 4]
```diff
@@ -388,3 +392,131 @@ env secret=apisix;
...
+            local values = secret.secrets()
+            ngx.say(#values)
+        }
+    }
+--- request
+GET /t
+--- response_body
+0
+--- error_log
+secret manager not exits
```
Comment (soulbird) [inline t/config-center-yaml/secret.t:444]: ditto

File: t/config-center-yaml/secret.t [hunk 5]
```diff
@@ -388,3 +392,131 @@ env secret=apisix;
...
+    }
+--- request
+GET /t
+--- response_body
+len: 1
+id: vault/1
+prefix: kv/apisix
+token: $ENV://VAULT_TOKEN
+uri: http://127.0.0.1:8200
+
```
Comment (soulbird) [inline t/config-center-yaml/secret.t:477]: There is no need to add this test

File: t/config-center-yaml/secret.t [hunk 6]
```diff
@@ -388,3 +392,131 @@ env secret=apisix;
...
+        }
+    }
+--- request
+GET /t
+--- response_body
+value
+
+
+
+=== TEST 21: secret.fetch_by_uri, no sub key value with the token in an env var
```
Comment (soulbird) [inline t/config-center-yaml/secret.t:503]: This test case already exists

File: t/plugin/basic-auth.t [hunk 7]
```diff
@@ -540,3 +544,73 @@ GET /echo
...
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- response_body
+passed
```
Comment (soulbird) [inline t/plugin/basic-auth.t:616]: You only configured the route, you also need to verify whether the route is effective

File: t/config-center-yaml/secret.t (hunk 3 above)
Comment (shreemaan-abhishek) [reply t/config-center-yaml/secret.t:419]: I just duplicated all the test cases containing the token with just one change: replace the token with the environment variable.

File: t/config-center-yaml/secret.t [hunk 8]
```diff
@@ -14,6 +14,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+BEGIN {
+    $ENV{VAULT_TOKEN} = "root";
```
Comment (soulbird) [inline t/config-center-yaml/secret.t:18]: Why was this added?

**Comment by soulbird** [comment]
IMO, just adding a test case for t/secret/vault.t is enough. The test cases under t/plugin seem redundant, don't you think?

**Comment by shreemaan-abhishek** [comment]
> The test cases under t/plugin seem redundant, don't you think?

I thought the same. But then I thought they would provide better safety from potential nasty bugs in the future.

**Comment by soulbird** [comment]
> > The test cases under t/plugin seem redundant, don't you think?
> 
> I thought the same. But then I thought they would provide better safety from potential nasty bugs in the future.

ok

**Review by soulbird — APPROVED** [review]

File: apisix/secret/vault.lua [hunk 9]
```diff
@@ -26,6 +26,7 @@ local norm_path = require("pl.path").normpath
 local sub        = core.string.sub
 local rfind_char = core.string.rfind_char
 
+local env        = require("apisix.core.env")
```
Comment (spacewander) [inline apisix/secret/vault.lua:29]: Can't we import it like `core.string.sub`?

**Review by spacewander — COMMENTED** [review]
Let's update https://github.com/apache/apisix/blob/master/docs/en/latest/terminology/secret.md

File: apisix/secret/vault.lua [hunk 10]
```diff
@@ -53,10 +54,15 @@ local function make_request_to_vault(conf, method, key, data)
     local req_addr = conf.uri .. norm_path("/v1/"
                 .. conf.prefix .. "/" .. key)
 
+    local token, _ = env.fetch_by_uri(conf.token)
```
Comment (soulbird) [inline apisix/secret/vault.lua:57]: Just use `core.env.fetch_by_uri`

Comment (shreemaan-abhishek) [reply apisix/secret/vault.lua:57]: I think it's okay to import using `require` and use because if in the future we need to use the env module again in this file then we won't have to re-import using `require`.

Also, at that time in the future, we will not have to replace `core.env.fetch_by_uri` with `env.fetch_by_uri`. WDYT?

**Review by soulbird — DISMISSED** [review]

File: apisix/secret/vault.lua [hunk 11]
```diff
@@ -26,6 +26,7 @@ local norm_path = require("pl.path").normpath
 local sub        = core.string.sub
 local rfind_char = core.string.rfind_char
```
Comment (spacewander) [inline apisix/secret/vault.lua:28]: We can avoid the blank line here? Other files don't add a blank line among localized variables.

**Review by spacewander — APPROVED** [review]

**Review by soulbird — APPROVED** [review]
